src/persist-client/src/internal/maintenance.rs

```diff
@@ -224,7 +231,13 @@ where
         V: Debug + Codec,
         D: Semigroup + Codec64 + Send + Sync,
     {
-        let _ = self.perform_in_background(machine, gc, compactor);
```

Comment:
ooooffff, so this was the bug? we'd normally have gotten a "must use" message from rustc about the Future, but we were explictly dong a `let _ =` which suppressed it. that's brutal

I'm already somewhat skeptical of `let _ =` (as evidenced by the `let () =` pattern, which admittedly has surprised other code readers), but this makes me wonder if we should at least have made this something like `let _: Vec<_> = ` before

Reply:
ooo interesting! https://rust-lang.github.io/rust-clippy/master/index.html#let_underscore_must_use

Reply:
Yeah, that's it!

> this makes me wonder if we should at least have made this something like `let _: Vec<_> =` before

This sounds plausible to me, though I'm not sure it would have stopped me from adding the bug... I'd missed the invariant that each of the futures in the Vec was running on a background task, so it didn't require the caller to drive it to completion. This is the reason that it was previously correct to ignore those futures, and I broke that invariant without realizing it was important here.

This is partly just me not fully internalizing the cancellation safety rules. I can also imagine adding a newtype wrapper for BoxFuture that _isn't_ `must_use` / which functions could return when they wanted to be explicit that it was safe to drop the return value? Unsure.

Reply:
👍